### ISSUE TITLE
Naming for TSE construction functions

### ISSUE THREAD

**Issue by antagomir** [microbiome/mia#45]
The `makeTreeSummarizedExperimentFromQIIME2` reads data from files, and can be considered an "import" function.

The other functions `makeTreeSummarizedExperimentFrom*` (* = DADA2, phyloseq, biom) convert R objects from other classes to TSE. 

I propose to separate the naming conventions so that conversion functions are called e.g. `convertphyloseqToTSE` and import functions are called e.g. `importQIIME2ToTSE`. 

Or we can have one wrapper, like: `convertToTSE(x)` or `makeTSE(x)` where the class of x can be automatically recognized from the list of accepted ones (DADA2, phyloseq, biom), or optionally, explicitly specified by the user if that is desired for some reason. This could simplify the procedures. I am not sure if we really need many different functions with long names, corresponding to conversions from different classes. 

Similar comment converning the import functions, we could just have `importTSE(...)`

**Comment by FelixErnst** [comment]
See #42 

TL;DR: origin and target class have to be in the function name to avoid any confusion. `make` vs. `convert` is an argument worth having, but from my point of view `make` wins 3:1 due to syntactic and history reasons.


From my point of view, the verbosity of the function name like `makeTreeSummarizedExperimentFrom*` is a plus. `convertToTSE` is just using a different verb, which is not that common in Bioconductor and omits the origin of the data.

If we try to break it down to its roots, a conversion effort it made up of three steps:
1. Check what type of object do I have? Lets call this class A
2. What is the function name for conversion from A to B?
3. Call the function

I would argue, the `make` verb makes step 2 a lot easier. With `convertToTSE` you cannot be sure, that your class of object is expected to work with this function. Only class B is given in the function name. So in this case you can go to step 3 and hope for the best or as a diligent user you refer to the man page. That takes you 2-3 min. From the developers point of view, this is an `if` call, which the user is capable to make, since he/she/they know what type of object is used. As a developer I can only guess and with a function name `convertToTSE` one could argue, that this claims to able to convert anything to a `TSE`, from a `logical` value to a single `integer`. This will cause issues, since conversion from `insert-fancy-new-object-type-I-have-never-heard-about` will not be covered by `convertToTSE`. I rather would have a `PR` adding a function, than an issue claiming that something is missing.

If the format `makeTreeSummarizedExperimentFrom*` is used, it is clear what is going on. We want to go to class B, so I expect a function name `makeB`. I know I have an object of class A, that results in me expecting `makeBFromA` as the function name. If that name does not exist, I know, I cannot convert it straight away. So step 1-3 are very easy to perform and, if you cannot find a function name in step 2 thats because it doesn't exist. It is a yes/no answer.

If the format `convertAToB*` is used, the order of what I have to look for is reversed. This is not a big deal and we might want to work with that. What speaks against this is, that every other function name in Bioconductor performing the same type of conversion I know of is using `make`. If we are getting really picky and compare it to `R` syntax, `make` is the equivalent to `b <- a`, whereas `convert` is `a -> b`. I know, that this is valid syntax, but I also know, that I feel queasy looking at it.

**Comment by antagomir** [comment]
All ok for `make` over `convert` and agree with most arguments.

Only point where I might like to see comments from others is the `makeBfromA` versus just having the wrapper`makeB` that works for many different objects. I can see the rationale in explicitly naming A and B here but if a user gives an invalid object that is not accepted, there can be an informative warning message that tells which types of objects are accepted and this could already help to avoid most issue reports.

**Comment by microsud** [comment]
Both of you are making valid points. If we think of the user, the `makeTSE()` wrapper with an argument for providing a `class = "phyloseq"` will still require user to type the format in an argument and again has to be sure with lower/upper case *"Phyloseq"* or *"phyloseq"*. But then, `makeTreeSummarizedExperimentFromPhyloseq`  is too long (40 characters, ideal coding line 80 characters limit). I think `makeTSEFromPhyloseq` covers two aspects, shorter/cleaner naming and is more user-friendly and informative, telling user clearly which inputs are supported by `mia`. Combined with the `tab` button and/or suggestions that pop up when writing in `Rstudio` this will help in a better user experience.

**Comment by antagomir** [comment]
Hi - in `makeTSE()` it is not necessary to write "phyloseq" as an argument (as you mention) if the function itself can recognize the type, which is straightfwd. It could be provided as an optional argument for users who want to explicitly specify those things. But it is not necessary and therefore not a full argument against `makeTSE`. 

I agree about the need to aim at shorter function names.

**Comment by microsud** [comment]
Okay. Like Felix mentioned, I also think that `makeTSE()` at first instance suggests unlimited options, and then as developer he needs to guess. I think its better to have separate functions which balances the needs of both the developer-maintainer and the user. Plus provides an opportunity for novice users to suggest potential enhancements clearly and raise issues to a specific function.

**Comment by FelixErnst** [comment]
I am not sure, what we are talking about. Just to understand it correctly, what is the aim of shortening the function name?
- make it easier for novice users
- make it easier to use in general 
- shorter function name for less typing

> I can see the rationale in explicitly naming A and B here but if a user gives an invalid object that is not accepted, there can be an informative warning message that tells which types of objects are accepted and this could already help to avoid most issue reports.

This would need to be an error, but if there is no function, there is no error, there is no issue. 100% no issue, without taking away anything, since it is not implemented either way.

Regarding implementation: How would you do it? S4-dispatch? Cannot do it, since you would require importing of all packages to be able to dispatch to a class (e.g. `phyloseq` would need to move from Suggest to Imports). A generic dispatch to `ANY`? Than it is not clear, which classes are allowed. No tooltips, not even in the man page. One would need to get an error with a message, to know which objects are allowed.

At this point this function would be user-facing only. In this situation it would be educationally beneficial to let the user figure out, what type of object they are using. If they don't know, they need to figure it out anyway, because how can anyone run an  analysis, if they don't know, what object the are using.

**Comment by antagomir** [comment]
Ok if implementation would lead to new technical problems with imports vs suggests, then I think it is easier to abandon the idea. In general, many functions in R accept multiple types. Anyway, there seems to be more support for specific functions for each conversion, so let's do that then.

Shortening the names: points 2-3 ie easier to use in general and faster to type.

Import functions: did we already reach an agreement to separate `make` and `import` functions?

**Comment by FelixErnst** [comment]
I don't have any `import*` functions in my sessions, so from my point of view, the same things as for `convert` apply. 

`makeTxDbFromGFF` is the example case from `GenomicFeatures`. The only `import*` function I can think off, is `import.gff` from `rtracklayer`, but that is a special case in my opinion, since all the imports handled like this, have a defined output type (In Bioconductor a `gff` will always be imported as a `GRanges` objects). 

Also this would only apply to the import of `biom` files, wouldn't it? The other `make*` functions apply to objects already available in the R session, for which the `import` verb might not be as fitting. `loadFromBiom` is already implemented. Should we rename that function?

Regarding point 2-3: I never type these long names, since with Tab I can directly choose in RStudio, RGui or even the console.

**Comment by antagomir** [comment]
Re: opening message in this issue thread:

The `makeTreeSummarizedExperimentFromQIIME2` _reads data from files,_ and can be considered an "import" function.

The other functions `makeTreeSummarizedExperimentFrom*` (* = DADA2, phyloseq, biom) _convert R objects from other classes_ to TSE.

I propose to separate the naming conventions so that conversion functions are called e.g. `makeTSEfromphyloseq` and import functions are called e.g. `importQIIME2ToTSE` or something.

**Comment by FelixErnst** [comment]
Sure, the QIIME results are saved as files. I missed that. Can you prepare a PR?

**Comment by antagomir** [comment]
Is `importQIIME2ToTreeSummarizedExperiment` or `importTreeSummarizedExperimentfromQIIME2` a good new name? 

Or could it be just `importQIIME2` ? (I like this one the best - not sure if TSE has to be explicitly mentioned here).

This function was contributed by @yiluheihei - I would like to hear what he thinks and if he could still make one more PR to update the name..? At least his approval/comment would be good.

**Comment by yiluheihei** [comment]
I think `import` is more convenient for QIIME2 results.`TreeSummarizedExperiment` is the main class in this package, so I like the name `importQIIME2` as antagomir.

**Comment by FelixErnst** [comment]
Apologies for being this picky.

I don't like the `import` prefix, since it suggests that the data is transformed. The prefix `load` is much more neutral, since it indicates, that the data is left untouched. 

For example `import.gff`, chops down the tabular data from a `gff` file, into seqnames, `IRanges` and a `DataFrame`, which columns are filled dynamically depending on which type of object makes most sense or is expected.

The functions we are using here, are just loading data and maybe include a bit of renaming. The dimensions stay almost the same for each part of the `QIIME2` dataset, so I personally don't regarding this as importing.

Having said this, I would prefer `loadFromQIIME2` as a wrapper for `makeTreeSummarizedExperimentFromQIIME2`, following the naming of `loadFromBiom`.

**Comment by antagomir** [comment]
Sounds fine to me. If @yiluheihei likes to make the PR it is very welcome, or otherwise I can make it (just changing the name of this function). Unless there are further suggestions.

**Comment by yiluheihei** [comment]
Okay, I will make a PR to rename it to `loadFromQIIME2()`.

**Comment by FelixErnst** [comment]
Would a wrapper be also executable? Have a look at `?loadFromBiom`

**Comment by antagomir** [comment]
Thanks @yiluheihei - !
